crates/wast/src/ast/types.rs

```diff
@@ -246,22 +252,102 @@ impl<'a> Parse<'a> for FunctionType<'a> {
     }
 }
 
+/// A struct type with fields.
+#[derive(Clone, Debug)]
+pub struct StructType<'a> {
+    /// The fields of the struct
+    pub fields: Vec<StructField<'a>>,
+}
+
+impl<'a> StructType<'a> {
+    fn finish_parse(&mut self, parser: Parser<'a>) -> Result<()> {
+        while parser.peek2::<kw::field>() {
+            parser.parens(|parser| {
+                self.fields.push(parser.parse()?);
+                Ok(())
+            })?;
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Parse<'a> for StructType<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        parser.parse::<kw::r#struct>()?;
+        let mut ret = StructType {
+            fields: Vec::new(),
+        };
+        ret.finish_parse(parser)?;
+        Ok(ret)
+    }
+}
+
+/// A field of a struct type.
+#[derive(Clone, Debug)]
+pub struct StructField<'a> {
+    /// An optional identifier for name resolution.
+    pub id: Option<ast::Id<'a>>,
+    /// Whether this field may be mutated or not.
+    pub mutable: bool,
+    /// The value type stored in this field.
+    pub ty: ValType<'a>,
+}
+
+impl<'a> Parse<'a> for StructField<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        parser.parse::<kw::field>()?;
+        let id = parser.parse()?;
+        let (ty, mutable) = if parser.peek2::<kw::r#mut>() {
+            let ty = parser.parens(|parser| {
+                parser.parse::<kw::r#mut>()?;
+                parser.parse::<ValType<'a>>()
```

Comment:
The type annotation of `ValType<'a>` can probably be removed here since it'll be inferred when this is stored inside of `StructField`